### ISSUE TITLE
support spaces embedded in tag values in Graphite plain-text ingestion

### ISSUE THREAD

**Issue by skeagy** [VictoriaMetrics/VictoriaMetrics#3102]
Need to support ingestion of graphite plain text protocol with spaces embedded in values. For example:
```
echo "foo.bar.baz;tag1=value1;description='my descr with spaces';tag3=val3 123 `date +%s`" | nc -n 127.0.0.1 2003
echo 'foo.bar.baz;tag1=value1;description=my\ descr\ with\ spaces;tag3=val3' 456 `date +%s` | nc -n 127.0.0.1 2003
```

Preferably, support delimiting with single or double quotes to suit different use cases (e.g. if the values have embedded single quotes or apostrophes, would be easier to use double-quotes as delimiter). Using a backslash to escape the space and treat it as a part of the tag value would also be acceptable. But that could break existing implementations if people use backslash as part of a tag or value. A way around this to please everyone might be to add a new flag to Victoriametrics such as -graphiteTagSupportSpaces. When this flag is set, the new handling behavior including single or double quotes or backslash to escape spaces is supported. If people have backwards compatibility issues, they can leave the flag unset (the default value) until they can migrate their data to not use embedded backslashes in the tags or values, or to escape the backslash they need to embed by putting another backslash in front of it.

The issue (without the suggested fix) was requested here: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/99

But it seems the fix: https://github.com/VictoriaMetrics/VictoriaMetrics/commit/9bc8484ab67accf3e480ab9f251f42f2314c0f99
was incomplete, by recognizing it as an exception and gracefully handling it, rather than actually supporting the embedded space(s). The main obstacle to full support seems to be ambiguity about interpreting white space. This concern is easily circumvented if tag values (and/or tag names) are delimited by single or double quotes, or if backslashes escape any spaces embedded in tag names or values.

Note that space is permitted character in label values from Graphite plaintext protocol, as discussed in the text:
https://graphite.readthedocs.io/en/latest/tags.html

> Tag values must also have a length >= 1, they may contain any ascii characters except ; and the first character must not be ~. UTF-8 characters may work for names and values, but they are not well tested and it is not recommended to use non-ascii characters in metric names or tags.

There is some interop/gray area because they do not provide an example of the plain text protocol using white space (e.g. tabs and spaces) in the tag names or values.

**Comment by skeagy** [comment]
Confirmed this is an issue in  build version: victoria-metrics-20220808-173243-tags-v1.80.0-0-gad00f4aaa. Got this error message:

2022-09-13T14:40:43.998Z        error   VictoriaMetrics/lib/protoparser/graphite/parser.go:160  cannot unmarshal Graphite line "foo_bar_baz;tag1='value1 with spaces';tag2=value2 123 1663080043": cannot unmarshal value from "with": unparsed tail left after parsing float64 from "with": "with"

**Comment by hagen1778** [comment]
cc @f41gh7 @valyala

**Comment by valyala** [comment]
@skeagy , thanks for the bugreport and for the proposed fix! The commit 41f8c2987ddc04bf1bf619937c72ac9dd7af68a9 changes the parsing logic for Graphite line protocol, so now it supports whitespace in metric names and tags out of the box according to [the spec](https://graphite.readthedocs.io/en/latest/tags.html) without the need to fall back to non-standard escaping and/or quoting.

The commit will be included in the next release of VictoriaMetrics.

**Comment by valyala** [comment]
VictoriaMetrics accepts whitespaces in tag names and tag values ingested via [Graphite plain-text protocol](https://docs.victoriametrics.com/#how-to-send-data-from-graphite-compatible-agents-such-as-statsd) starting from the [release v1.82.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing this feature request as done.
